pennylane/devices/qubit/measure.py

```diff
@@ -108,6 +108,28 @@ def csr_dot_products(
     return math.real(math.squeeze(res))
 
 
+def full_dot_products(
```

Comment:
@albi3ro you can probably gain further in performance by exploiting the structure of the hermitian
```
cicj = np.array([state.conj()[i]*conj[j] for i in range(np.shape(h)[0]) for j in range(np.shape(h)[0]) if i<j])
result = np.einsum('i,i->', np.diag(h), np.abs(state)**2) + 2*np.real(np.eisnum('i,i->', cicj, h.T[np.tril_indices(np.shape(h)[0],-1)])

```
```apply_operation``` will probably need to take this into account
https://github.com/PennyLaneAI/pennylane/blob/77f262d57a857687ea510e14c82ef0cbd4862057/pennylane/devices/qubit/apply_operation.py#L60 
https://github.com/PennyLaneAI/pennylane/blob/77f262d57a857687ea510e14c82ef0cbd4862057/pennylane/devices/qubit/apply_operation.py#L103

Reply:
This could potentially be a follow up PR. Something like "Register specialized kernel for hermitians to `apply_operation`".